### PR TITLE
feat: add std/expect

### PR DESCRIPTION
**Pull request by kt3k**
Continued from #3627

This PR adds `std/expect` mostly compatible with `expect` of jest.

Remaining:
- [x] toHaveLength
- [x] toHaveProperty
- [x] toContain
- [x] toContainEqual
- [x] Fix toBeCloseTo tolerance handling

Mock related:
- [x] fn (mock)
- [x] toHaveBeenCalled
- [x] toHaveBeenCalledTimes
- [x] toHaveBeenCalledWith
- [x] toHaveBeenLastCalledWith
- [x] toHaveBeenNthCalledWith
- [x] toHaveReturned
- [x] toHaveReturnedTimes
- [x] toHaveReturnedWith
- [x] toHaveLastReturnedWith
- [x] toHaveNthReturnedWith

Testing:
- [x] add test cases

We'd avoid supporting snapshot related APIs for first pass (`toMatchSnapshot` `toMatchInlineSnapshot` `toThrowErrorMatchingSnapshot` `toThrowErrorMatchingInlineSnapshot`)

Done:
- [x] documentation

closes #3813 1779

### CODE REVIEW

**Comment by andrewthauer** [comment]
Very much looking forward to this. Curious though, would it make sense to put this under `std/testing` or may `std/testing/expect` instead?

**Comment by iuioiua** [comment]
We thought the `expect()` implementation, once complete with testing, would be of a sufficient size that it might justify having it as its own sub-module. Plus, it'd be more visible.

**Comment by kt3k** [comment]
> Curious though, would it make sense to put this under std/testing or may std/testing/expect instead?

I'm in favor of keeping the implementation in `std/expect`, but I'm open to the idea of re-exporting this from somewhere in `std/testing`, such as from `std/testing/bdd.ts`.

**Comment by andrewthauer** [comment]
Out of curiosity, is the plan to keep all modules at the top level or will there be any nesting?

I personally think I'd find it more intuitive to scan top level areas first, then drill into them further. For example I'd find `testing` and then look inside to see `bdd` and `expect`, etc.. A possible structure might look like:

```
└── testing
    ├── bdd
    │   └── mod.ts
    ├── expect
    │   └── mod.ts
    └── mod.ts

```

**Comment by kt3k** [comment]
We generally prefer top level modules over nested structure. For example, we moved some complex modules from `encoding/` to top-level to avoid the nested structure in the past. cf https://github.com/denoland/deno_std/issues/3123

**Comment by andrewthauer** [comment]
Right. For encoding that makes sense to me. I personally wouldn't think to look under that initially for json,yaml, etc.

I would say that `expect` is much less clear to me own its own at the top level then the encoding ones. Putting it under testing would give it context and make it more clear. My 2 cents anyways. I'll will say no more now 😀.

File: expect/_to_be.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import { assertNotStrictEquals } from "../assert/assert_not_strict_equals.ts";
+import { assertStrictEquals } from "../assert/assert_strict_equals.ts";
+
+export function toBe(context: MatcherContext, expect: unknown): MatchResult {
+  if (context.isNot) {
+    assertNotStrictEquals(context.value, expect, context.customMessage);
+  } else {
+    assertStrictEquals(context.value, expect, context.customMessage);
+  }
+}
```
Comment (bartlomieju) [inline expect/_to_be.ts:13]: What's the reason for splitting into so many small modules?

File: expect/_to_be_close_to.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+import { MatcherContext, MatchResult } from "./_types.ts";
+import { AssertionError } from "../assert/assertion_error.ts";
+
+// TODO(kt3k): tolerance handling is wrong
```
Comment (bartlomieju) [inline expect/_to_be_close_to.ts:6]: Is there a test that covers this tolerance?

**Review by bartlomieju — COMMENTED** [review]
This looks mostly good to me, I'm a bit worried about a number of modules that were added - if you're unit testing a file that has no dependencies besides pulling `expect/mod.ts` then the overhead of loading all these small modules might be non-trivial. Could you reconsider adding a single module that has all the matchers inside it?

File: expect/_to_be_close_to.ts (hunk 2 above)
Comment (kt3k) [reply expect/_to_be_close_to.ts]: Oops, this has been already resolved in https://github.com/denoland/deno_std/pull/3814/commits/357ecb3f2cd0f2ab9a2748cc7793c6f9f8f3c1ef

**Comment by kt3k** [comment]
> if you're unit testing a file that has no dependencies besides pulling expect/mod.ts then the overhead of loading all these small modules might be non-trivial.

Fair enough. I'll move all matchers into `_matchers.ts`

**Review by bartlomieju — APPROVED** [review]

**Comment by kt3k** [comment]
Landing this as `std/expect` for now. Let's revisit `std/expect` vs `std/testing/expect` before stabilizing this.

**Pull request merged**